File: Day_2/Red-Nosed_Reports.go
package main

import (
	"fmt"
	// "sort"
	"strings"
	"os"
	// "reflect"
	// "math"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

var print = fmt.Println

// func checkList(l []int) bool {
// 	sorted_as := make([]int, len(l))
// 	sorted_ds := make([]int, len(l))
// 	copy(sorted_as, l)
// 	copy(sorted_ds, l)
// 	sort.Ints(sorted_as)
// 	sort.Sort(sort.Reverse(sort.IntSlice(sorted_ds)))
// 	if reflect.DeepEqual(sorted_as, l) || reflect.DeepEqual(sorted_ds, l){
// 		for i:=1; i<len(l); i++{
// 			t := int64(math.Abs(float64(l[i-1]-l[i])))
// 			if !(t==1 || t==2 || t==3){
// 				return false
// 			}
// 		}
// 		return true
// 	}
// 	return false
// }

func checkList(l []int) bool {
	isAscending, isDescending := true, true

	for i := 1; i < len(l); i++ {
		diff := l[i] - l[i-1]
		if diff < -3 || diff > 3 || diff == 0 {
			return false
		}
		isAscending = isAscending && (diff > 0)
		isDescending = isDescending && (diff < 0)
	}
	return isAscending || isDescending
}

func remove_and_check(l []int) bool {
	for i:=0; i< len(l);i++{
		temp := make([]int, len(l)-1)
		copy(temp[:i], l[:i])
		copy(temp[i:], l[i+1:])
		if checkList(temp){
			return true
		}
	}
	return false
}



func main(){
	// Part 1 and Part 2
	dat, err := os.ReadFile("D2.txt")
	check(err)
	safe := 0
	data := strings.Split(strings.ReplaceAll(string(dat), "\r\n", "\n"), "\n")
	for _, line := range data{
		split_line := strings.Fields(line)
		lis := make([]int, len(split_line))
		for i, item := range(split_line){
			fmt.Sscanf(item, "%d", &lis[i])
		}
		if checkList(lis) || remove_and_check(lis){ // remove the 2nd or condition for part 1 answer
			safe+=1
		}
	}
	print(safe)
}

File: Day_2/Red-Nosed_Reports.py
# Part 1
with open("D2.txt", "r") as f:
    data = f.read().splitlines()
    safe = 0
    for i in data:
        c=0
        l = list(map(int,i.split()))
        if l == sorted(l) or l == sorted(l, reverse=True):
            for j in range(1, len(l)):
                temp = abs(l[j] - l[j-1])
                if not (temp ==1 or temp==2 or temp==3):
                    c=1
                    break
        else:
            continue
            
        if c == 0:
            safe += 1
    print(safe)
    # 564

# Part 2
def check_list(l):
    c=0
    if l == sorted(l) or l == sorted(l, reverse=True):
            for j in range(1, len(l)):
                temp = abs(l[j] - l[j-1])
                if not (temp ==1 or temp==2 or temp==3):
                    c=1
                    break
    else :
        return False
    if c == 0:
        return True
    return False

def remove_and_check(l):
    for i in range(len(l)):
        temp = l[i]
        l.pop(i)
        if check_list(l):
            return True
        l.insert(i, temp)
    return False        



with open("D2.txt", "r") as f:
    data = f.read().splitlines()
    safe = 0
    for i in data:
        c=0
        l = list(map(int,i.split()))
        if check_list(l) or remove_and_check(l):
            safe += 1
    print(safe)
    # 604